packages/product/src/services/product-module-service.ts

```diff
@@ -369,6 +375,145 @@ export default class ProductModuleService<
+              productData.thumbnail = isString(productData.images[0])
+                ? (productData.images[0] as string)
+                : (productData.images[0] as { url: string }).url
+            }
+
+            if (productData.is_giftcard) {
+              productData.discountable = false
+            }
+
+            if (productData.images?.length) {
+              productData.images = await this.productImageService_.upsert(
+                productData.images.map((image) =>
+                  isString(image) ? image : image.url
+                ),
+                sharedContext
+              )
+            }
+
+            if (productData.tags?.length) {
+              productData.tags = await this.productTagService_.upsert(
+                productData.tags,
+                sharedContext
+              )
+            }
+
+            if (isDefined(productData.type)) {
+              const productType = (
+                await this.productTypeService_.upsert(
+                  [productData.type as ProductTypes.CreateProductTypeDTO],
+                  sharedContext
+                )
+              )
+
+              productData.type = productType?.[0]
+            }
+
+            return productData as ProductServiceTypes.UpdateProductDTO
+          })
+        )
+
+        // TODO
+        // Shipping profile is not part of the module
+        // as well as sales channel
+
+        const products = await this.productService_.update(
+          productsData,
+          sharedContext
+        )
+
+        const productByIdMap = new Map<string, TProduct>(
+          products.map((product) => [product.id, product])
+        )
+
+        const productOptionsData = [...productOptionsMap]
+          .map(([id, options]) => {
+            return options.map((option) => {
+              return {
+                ...option,
+                product: productByIdMap.get(id)!,
+              }
+            })
+          })
+          .flat()
+
+        const productOptions = await this.productOptionService_.create(
+          productOptionsData,
+          sharedContext
+        )
+
+        for (const variants of productVariantsMap.values()) {
+          variants.forEach((variant) => {
+            variant.options = variant.options?.map((option, index) => {
+              const productOption = productOptions[index]
+              return {
+                option: productOption,
+                value: option.value,
+              }
+            })
+          })
+        }
+
+        const variants = await Promise.all(
+          [...productVariantsMap].map(async ([id, variants]) => {
+            return await this.productVariantService_.create(
+              productByIdMap.get(id)!,
+              variants as unknown as ProductTypes.CreateProductVariantOnlyDTO[],
+              sharedContext
+            )
+          })
+        )
+
+        return products
+      },
+      { transaction: sharedContext?.transactionManager }
```

Comment:
todo: i think we are missing some steps, we are not updating existing variants, deleting the variants that needs to be deleted for example. For the deletion we could argue that the users could call another method to delete the variants from a product but we can discuss that

Reply:
I'm gonna do this in a follow up PR, this needs more love and tests. It started getting real messy real quick.

Reply:
Can you rename the pr as a partial update so far please :)

Reply:
@adrien2p put it in the same PR